src/terminal/parser/stateMachine.cpp

```diff
@@ -1973,125 +2034,161 @@ bool StateMachine::FlushToTerminal()
                 }
-                _EnterGround();
-                // No need to cache the run, since we've dealt with it now.
-                cacheUnusedRun = false;
+            }
+
+            // The only way we get here is if the above loop ran into the ground state.
+            // This means we're starting a new "chunk".
+            // --> Reset beg(in).
+            _runBeg = _runEnd;
+
+            // Ran out of text. Return.
+            if (_runBeg >= _currentString.size())
+            {
+                return;
+            }
+
+            // Plain text? Go back to the start, where we process plain text.
+            if (!Utils::IsActionableFromGround(til::at(_currentString, _runEnd)))
+            {
+                break;
             }
         }
-        else if (_state == VTStates::SosPmApcString || _state == VTStates::DcsPassThrough || _state == VTStates::DcsIgnore)
-        {
-            // There is no need to cache the run if we've reached one of the
-            // string processing states in the output engine, since that data
-            // will be dealt with as soon as it is received.
-            cacheUnusedRun = false;
-        }
+    }
+}
 
-        // If the run hasn't been dealt with in one of the cases above, we cache
-        // the partial sequence in case we have to flush the whole thing later.
-        if (cacheUnusedRun)
+void Microsoft::Console::VirtualTerminal::StateMachine::_processStringIncompleteSequence()
+{
+    const auto run = _CurrentRun();
+    auto cacheUnusedRun = true;
+
+    // One of the "weird things" in VT input is the case of something like
+    // <kbd>alt+[</kbd>. In VT, that's encoded as `\x1b[`. However, that's
+    // also the start of a CSI, and could be the start of a longer sequence,
+    // there's no way to know for sure. For an <kbd>alt+[</kbd> keypress,
+    // the parser originally would just sit in the `CsiEntry` state after
+    // processing it, which would pollute the following keypress (e.g.
+    // <kbd>alt+[</kbd>, <kbd>A</kbd> would be processed like `\x1b[A`,
+    // which is _wrong_).
+    //
+    // At the same time, input may be broken up arbitrarily, depending on the pipe's
+    // buffer size, our read-buffer size, the sender's write-buffer size, and more.
+    // In fact, with the current WSL, input is broken up in 16 byte chunks (Why? :(),
+    // which breaks up many of our longer sequences, like our Win32InputMode ones.
+    //
+    // As a heuristic, this code specifically checks for a trailing Esc or Alt+key.
+    // If we encountered a win32-input-mode sequence before, we know that our \x1b[?9001h
+    // request to enable them was successful. While a client may still send \x1b{some char}
+    // intentionally, it's far more likely now that we're looking at a broken up sequence.
+    // The most common win32-input-mode is ConPTY itself after all, and we never emit
+    // \x1b{some char} once it's enabled.
+    if (_isEngineForInput)

```

Comment:
in my work on OSC sequences i realized... I don't think caching actually does anything for the output engine any longer. The cached sequence is never emitted because OutputStateMachineEngine::ActionPassThroughString does nothing.

Which makes sense, actually. Since ConPTY became passthrough mode, there's no reason to cache. We just yeet everything through.